src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/ParameterCapturingService.cs

```diff
@@ -1,81 +1,270 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using Microsoft.Diagnostics.Monitoring.HostingStartup.ParameterCapturing.Eventing;
 using Microsoft.Diagnostics.Monitoring.HostingStartup.ParameterCapturing.FunctionProbes;
 using Microsoft.Diagnostics.Monitoring.StartupHook;
+using Microsoft.Diagnostics.Monitoring.StartupHook.MonitorMessageDispatcher.Models;
+using Microsoft.Diagnostics.Tools.Monitor.ParameterCapturing;
+using Microsoft.Diagnostics.Tools.Monitor.StartupHook;
 using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Extensions.Hosting;
 using Microsoft.Extensions.Logging;
 using System;
 using System.Collections.Generic;
+using System.Diagnostics;
 using System.Reflection;
 using System.Threading;
+using System.Threading.Channels;
 using System.Threading.Tasks;
 
 namespace Microsoft.Diagnostics.Monitoring.HostingStartup.ParameterCapturing
 {
     internal sealed class ParameterCapturingService : BackgroundService, IDisposable
     {
         private long _disposedState;
-        private readonly bool _isAvailable;
+
+        private ParameterCapturingEvents.ServiceNotAvailableReason? _notAvailableReason;
+        private string _notAvailableDetails = string.Empty;
 
         private readonly FunctionProbesManager? _probeManager;
+        private readonly ParameterCapturingEventSource _eventSource = new();
         private readonly ILogger? _logger;
 
+        private Channel<StartCapturingParametersPayload>? _requests;
```

Comment:
If the design of this feature is that there is only one capturing session ever enabled at a time, then I think there should only be one channel that contains both the start and stop requests. Or maybe it just the start requests, but the request object has a completion source that can be signaled by the StopCapture method and the ExecuteAsync will just pop the request of the channel, start it, and wait for the completion source to be signaled before stopping it.

Reply:
No, but you can put it on the payload itself so the completion is only associated with that exact request.

Reply:
Another thing to consider is that you could add a capture ID property to both the start and stop requests. In the hosting startup, you can then have a dictionary of that ID to the request payload object; on a start request, add it to the dictionary, on a stop request, signal the completion source and remove it from the dictionary. This would absolve the need to have a notion of a singleton request except for the processing routine which would still only handle one request at a time.